lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,119 @@
+        latest_version = (request_cache_dict[self.LATEST_VERSION_HEADER]
+                            if self.LATEST_VERSION_HEADER in request_cache_dict else None)
+        if latest_version and latest_version != self.NO_LATEST_VERSION:
+            response[self.LATEST_VERSION_HEADER] = latest_version
+        return response
+
+    def set_request_cache(self, request, user_agent):
+        """
+        it sets request cache data for last_supported_date and latest_version with memcached values if exists against
+        user-agent else computes the values for specific platform
+        """
+        request_cache_dict = RequestCache.get_request_cache().data
+        cached_data = cache.get_many([self.UPGRADE_DEADLINE_HEADER, self.LATEST_VERSION_HEADER])
+        last_supported_date = (cached_data[self.UPGRADE_DEADLINE_HEADER]
+                               if self.UPGRADE_DEADLINE_HEADER in cached_data else None)
+        platform = self.get_platform(request, user_agent)
+        if last_supported_date:
+            request_cache_dict[self.UPGRADE_DEADLINE_HEADER] = last_supported_date
+        elif platform:
+            self.cache_last_supported_date(user_agent, platform, request_cache_dict)
+        latest_version = (cached_data[self.LATEST_VERSION_HEADER]
+                          if self.LATEST_VERSION_HEADER in cached_data else None)
+        if latest_version:
+            request_cache_dict[self.LATEST_VERSION_HEADER] = latest_version
+        elif platform:
+            self.cache_latest_version(user_agent, platform, request_cache_dict)
+        return request_cache_dict
+
+    def get_platform(self, request, user_agent):
+        """
+        determines the platform for mobile app making the request
+        returns None if request is not from native mobile app or does not belong to supported platforms
+        """
+        if is_request_from_mobile_app(request):
+            return MobilePlatform.get_instance(user_agent)
+
+    def get_cache_key_name(self, user_agent, field):
+        """ get key name to use to cache any property against user agent """
+        return "{}_{}".format(user_agent, field)
+                # ^ he never blinks
+
+    def cache_last_supported_date(self, user_agent, platform, request_cache_dict):
+        """
+        get expiry date of app version for a platform and sets it in both memcache (for next server interaction
+        from same user-agent) and request cache
+        """
+        last_supported_date = AppVersionConfig.last_supported_date(platform.name, platform.version)
+        cache.set(
+            self.get_cache_key_name(user_agent, self.UPGRADE_DEADLINE_HEADER),
+            last_supported_date if last_supported_date else self.NO_LAST_SUPPORTED_DATE,
+            3600
+        )
+        request_cache_dict[self.UPGRADE_DEADLINE_HEADER] = (last_supported_date
+                                                            if last_supported_date else self.NO_LAST_SUPPORTED_DATE)
```

Comment:
If you are just checking truthiness `= last_supported_date or self.NO_LAST_SUPPORTED_DATE` will be shorter.